### PR TITLE
[WIP][GLUTEN-5471][VL]feat: Support read Hudi COW table

### LINKED ISSUE #5471: [VL] Support read Hudi COW
### Description

Support read Hudi COW tables.

### PR DESCRIPTION
**Pull request by xushiyan**
## What changes were proposed in this pull request?

Support read hudi COW tables. Fixes #5471

## How was this patch tested?

Unit tests, integration tests.

### CODE REVIEW

File: gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    optionalNumCoalescedBuckets,
+    dataFilters,
+    tableIdentifier,
+    disableBucketedScan
+  ) {
+
+  override lazy val fileFormat: ReadFileFormat = ReadFileFormat.ParquetReadFormat
+
+  override protected def doValidateInternal(): ValidationResult = {
+    if (!requiredSchema.fields.exists(_.name == "_hoodie_record_key")) {
```
Comment (leesf) [inline gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformer.scala:55]: the check is not needed if users do not set meta columns when writing to hudi and we still could use velox to read parquet file?

Comment (yma11) [reply gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformer.scala:55]: +1. Another point is that we should also ensure MOR also fallback correctly as only COW is supported. Refer [code](https://github.com/apache/incubator-gluten/blob/main/gluten-delta/src/main/scala/org/apache/gluten/execution/DeltaScanTransformer.scala#L56-L60).

**Review by YannByron — APPROVED** [review]

File: gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformer.scala (hunk 1 above)
Comment (leesf) [reply gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformer.scala:55]: MOR table with ReadOptimized could be also supported.

Comment (xushiyan) [reply gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformer.scala:55]: agreed. thanks @leesf @yma11 for the comments

File: gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformerProvider.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * limitations under the License.
+ */
+package org.apache.gluten.execution
+
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.execution.FileSourceScanExec
+
+class HudiScanTransformerProvider extends DataSourceScanTransformerRegister {
+
+  override val scanClassName: String = "org.apache.hudi.HoodieFileIndex"
```
Comment (stream2000) [inline gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformerProvider.scala:24]: Should we change the `scanClassNam` from `file index` to `file format` according to https://github.com/apache/incubator-gluten/pull/3955/files#diff-7b2924f5d19292df8ae434c562c54924ea90de8a7be38669570cf7cb4d13402dR47

Comment (xushiyan) [reply gluten-hudi/src/main/scala/org/apache/gluten/execution/HudiScanTransformerProvider.scala:24]: it's actually better to stick to what the trait specifies for the variable name. See docs of `org.apache.gluten.execution.DataSourceScanTransformerRegister#scanClassName`
 
>    * For DataSource V1, it should be the child class name of
>   * [[org.apache.spark.sql.execution.datasources.FileIndex]].
>   *
>   * For DataSource V2, it should be the child class name of
>   * [[org.apache.spark.sql.connector.read.Scan]].

**Comment by xushiyan** [comment]
@yma11 @stream2000 @leesf if any of you're interested and have bandwidth, would you like to take over this PR (or open new one) and move forward? I'm bit short on availability.

**Comment by yma11** [comment]
> @yma11 @stream2000 @leesf if any of you're interested and have bandwidth, would you like to take over this PR (or open new one) and move forward? I'm bit short on availability.

@xushiyan Thanks for your changes. I will address the comments based on your PR.

**Comment by xushiyan** [comment]
> > @yma11 @stream2000 @leesf if any of you're interested and have bandwidth, would you like to take over this PR (or open new one) and move forward? I'm bit short on availability.
> 
> @xushiyan Thanks for your changes. I will address the comments based on your PR.

@yma11 thanks. any plan on your side to make change and get this merged?

**Comment by yma11** [comment]
> > > @yma11 @stream2000 @leesf if any of you're interested and have bandwidth, would you like to take over this PR (or open new one) and move forward? I'm bit short on availability.
> > 
> > 
> > @xushiyan Thanks for your changes. I will address the comments based on your PR.
> 
> @yma11 thanks. any plan on your side to make change and get this merged?

I will do it next week. Thanks.

**Pull request closed without merging**